### PR TITLE
[POC] enable trtllm fp4 from trtllm wheel

### PR DESCRIPTION
**Pull request by Alcanderian**
## Motivation

ATTENTION: flashinfer cutlass moe may not work after installing trtllm wheel

UPD 072025, archieve up to 134 tps with tp8

UPD 070525, archieve up to 128 tps
```
SGLANG_TRTLLM_GEN_MOE_EP_SIZE=2 SGLANG_ENABLE_TRTLLM_GEN_MOE=1 python3 -m sglang.launch_server \
--model-path /dev/shm/DeepSeek-R1-FP4 --trust-remote-code --quantization modelopt_fp4 --tp 8 \
--enable-flashinfer-moe --enable-ep-moe --enable-flashinfer-allreduce-fusion
```

ENV
```
pip3 install --no-cache-dir tensorrt-llm==1.0.0rc0 --no-deps
pip3 install --no-cache-dir tensorrt~=10.11.0 \
  tensorrt_cu12_bindings~=10.11.0 tensorrt_cu12~=10.11.0 \
  tensorrt_cu12_libs~=10.11.0 --no-deps
pip3 install --no-cache-dir nvtx mpi4py onnx onnx_graphsurgeon\>=0.5.2 \
  StrEnum accelerate\>=0.25.0 nvidia-modelopt\[torch\]~=0.31.0
pip3 install --no-cache-dir aenum backoff click_option_group \
  colored diffusers\>=0.27.0 etcd3 evaluate meson openai opencv-python-headless
pip3 install --no-cache-dir optimum ordered-set peft polygraphy \
  tornado\>=6.5.0 setuptools\<80 h5py==3.12.1 blake3

pip3 install -e "python[blackwell]"
pip3 install --no-cache-dir nvidia-nccl-cu12==2.27.6 --force-reinstall --no-deps
```

RUN
```
SGLANG_ENABLE_TRTLLM_GEN_MOE=1 python3 -m sglang.launch_server --model-path /dev/shm/DeepSeek-R1-FP4 \
    --trust-remote-code --quantization modelopt_fp4 --tp 8 --enable-flashinfer-moe

# ACC
python3 benchmark/gsm8k/bench_sglang.py --parallel 1400 --num-questions 1400
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:40<00:00, 32.91it/s]
Accuracy: 0.951
Invalid: 0.000
Latency: 40.467 s
Output throughput: 3108.238 token/s

lm_eval
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9659|±  | 0.005|
|     |       |strict-match    |     5|exact_match|↑  |0.9492|±  | 0.006|

# TPS
[2025-07-02 11:51:13 TP0] Decode batch. #running-req: 1, #token: 1236, token usage: 0.00, cuda graph: True, gen throughput (token/s): 119.54, #queue-req:
```

TODO: add a custom cublas_mm kernel to let router gemm always give fp32 output, https://github.com/sgl-project/sglang/pull/7711/files#diff-5b9e34dd492bd8a14702a18b594721091092276fad1cf8736fba6ef1f33c1b04R247

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by trevor-m** [comment]
Nice work! FYI the TRTLLM Gen MoE kernel is being added to flashinfer, so that will help to ease the deps setup.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 1]
```diff
@@ -56,6 +57,8 @@
     from aiter.fused_moe_bf16_asm import ck_moe_2stages
     from aiter.ops.shuffle import shuffle_weight
 
+TRTLMM_GEN_MOE_EP_SIZE = get_int_env_var("SGLANG_TRTLMM_GEN_MOE_EP_SIZE", "0")
```
Comment (hlu1) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]:
```suggestion
TRTLMM_GEN_MOE_EP_SIZE = get_int_env_var("SGLANG_TRTLLM_GEN_MOE_EP_SIZE", "0")
```

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 2]
```diff
@@ -353,6 +366,13 @@ class ModelOptFp4LinearMethod(LinearMethodBase):
     def __init__(self, quant_config: ModelOptFp4Config):
         self.quant_config = quant_config
 
+        if ENABLE_TRTLMM_GEN_MOE:
+            self.kernel = torch.ops.trtllm.nvfp4_gemm
```
Comment (hlu1) [inline python/sglang/srt/layers/quantization/modelopt_quant.py:569]: The trtllm nvfp4_gemm needs profiling to find the best config. How is that handled in sglang?

File: python/sglang/srt/models/deepseek_v2.py [hunk 3]
```diff
@@ -236,6 +238,12 @@ def forward(self, hidden_states):
             )
 
         # NOTE: For some unknown reason, router_gemm seems degrade accept length.
+        if ENABLE_TRTLMM_GEN_MOE and not self.is_nextn:
+            return torch.ops.trtllm.dsv3_router_gemm_op(
```
Comment (hlu1) [inline python/sglang/srt/models/deepseek_v2.py:253]: How is this one different from the `dsv3_router_gemm` on line 255?

File: python/sglang/srt/models/deepseek_v2.py [hunk 4]
```diff
@@ -425,14 +432,17 @@ def forward(
...
+            # router_logits: (num_tokens, n_experts)
+            router_logits = self.gate(hidden_states)
 
         current_stream = torch.cuda.current_stream()
         self.alt_stream.wait_stream(current_stream)
         shared_output = self._forward_shared_experts(hidden_states)
 
         with torch.cuda.stream(self.alt_stream):
+            if ENABLE_TRTLMM_GEN_MOE:
+                router_logits = self.gate(hidden_states)
```
Comment (hlu1) [inline python/sglang/srt/models/deepseek_v2.py]: Why this change?

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 5]
```diff
@@ -780,6 +1000,67 @@ def apply(
...
+                    scale_factor_use_ue8m0,
+                    is_scale_factor_swizzled,
+                )
+            )
+
+            # FIXME: tile_tokens_dim is hardcoded for now
+            tile_tokens_dim = 8
+
+            # https://github.com/NVIDIA/TensorRT-LLM/blob/v1.0.0rc1/tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py#L195
+            outputs = torch.ops.trtllm.fp4_block_scale_moe_runner(
```
Comment (hlu1) [inline python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: I don't know if you have benchmarked the prefill perf separately. trtllm gen MoE is optimized for decoding, i.e., small input num_tokens.

**Comment by azhurkevich** [comment]
based tbh

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 2 above)
Comment (azhurkevich) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:569]: There is probably none from what I can see. By default it'll use some heuristics, but they are not the best. There is autotuner in trtllm that handles this https://github.com/NVIDIA/TensorRT-LLM/pull/5207/files. I can bring that autotuner with flashninfer integration https://github.com/flashinfer-ai/flashinfer/pull/1214

Comment (Alcanderian) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:569]: We have no auto tunner in sglang for now

File: python/sglang/srt/models/deepseek_v2.py (hunk 3 above)
Comment (Alcanderian) [reply python/sglang/srt/models/deepseek_v2.py:253]: The dsv3_router_gemm below do not have cublas fallback, and F.linear do not support fp16 input with fp32 output for now.

File: python/sglang/srt/models/deepseek_v2.py (hunk 4 above)
Comment (Alcanderian) [reply python/sglang/srt/models/deepseek_v2.py]: This change is to reproduce trtllm profile's timeline

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 5 above)
Comment (Alcanderian) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: I have not done the benchmark on prefill, but I can see that it aslo be called in trtllm prefill stage. And by the way are there any fp4_gemm interface can apply the same weight/scaling factor layout as trtllm gen moe? Because we cannot change the layout of weight/scaling factor in the forward stage.

**Comment by azhurkevich** [comment]
fyi, we merged https://github.com/flashinfer-ai/flashinfer/pull/1214 and the plan is to enable these kernels through flashinfer in SGL

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 5 above)
Comment (hlu1) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: Yes, I believe trtllm has added it: https://github.com/NVIDIA/TensorRT-LLM/blob/d4d21a106e8176bf20e627ee432cca5ef920c325/tests/unittest/_torch/thop/test_fp4_gemm_quantize.py#L122
But the weight layout is different from MoE.
Are you planning to run MoE as individual gemms in the forward stage? If so, this might be close to what you want:
https://github.com/NVIDIA/TensorRT-LLM/blob/d4d21a106e8176bf20e627ee432cca5ef920c325/tensorrt_llm/_torch/models/modeling_llama_min_latency.py#L47
The precision is not nvfp4.

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 2 above)
Comment (hlu1) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:569]: Then the performance may not be optimal without profiling.

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 5 above)
Comment (Alcanderian) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: These two seem unable to provide any effective assistance.

Comment (hlu1) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: What I wanted to say is that the layout changes will not be compatible with single gemms, because some of the layout shuffling is for swiglu fusion. The only possibility is the nvfp4 version of GatedMLP kernels, which shares the same weight layout as MoE. These kernels can be built. They are just not checked into trtllm.

Comment (hlu1) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:1213]: Actually now I think about it, the GatedMLP kernels would be optimized for low latency too.
